objects/src/chain/mod.rs

```diff
@@ -1,39 +1,79 @@
-use super::{crypto::merkle::Mmr, AdviceInputsBuilder, ToAdviceInputs};
+use super::{
+    crypto::merkle::{MmrPeaks, PartialMmr},
+    utils::collections::{BTreeMap, Vec},
+    AdviceInputsBuilder, Digest, ToAdviceInputs,
+};
 
-// TODO: Consider using a PartialMmr that only contains the Mmr nodes that are relevant to the
-// transaction being processed.
+// CHAIN MMR
+// ================================================================================================
 
-/// A struct that represents the chain Mmr accumulator.
-/// This wraps the [Mmr] object and provides a simple interface to access / modify it.
-/// We use a custom type here as the traits we implement on this type could be context specific.
+/// A struct that represents the chain Merkle Mountain Range (MMR).
 ///
-/// The Mmr allows for efficient authentication of consumed notes during transaction execution.
-/// Authentication is achieved by providing an inclusion proof for the consumed notes in the
-/// transaction against the chain Mmr root associated with the latest block known at the time
+/// The MMR allows for efficient authentication of input notes during transaction execution.
+/// Authentication is achieved by providing inclusion proofs for the notes consumed in the
+/// transaction against the chain MMR root associated with the latest block known at the time
 /// of transaction execution.
-#[derive(Clone, Default, Debug)]
-#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
-pub struct ChainMmr(Mmr);
+///
+/// [ChainMmr] represents a partial view into the actual MMR and contains authentication paths
+/// for a limited set of blocks. The intent is to include only the blocks relevant for execution
+/// of a specific transaction (i.e., the blocks corresponding to all input notes).
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ChainMmr {
+    /// Partial view of the Chain MMR with authentication paths for the blocks listed below.
+    mmr: PartialMmr,
+    /// A list of `(block_num, block_hash)` tuples for all blocks for which the partial MMR
+    /// contains authentication paths.
+    blocks: Vec<(usize, Digest)>,
+}
 
 impl ChainMmr {
-    /// Returns a reference to the Mmr.
-    pub fn mmr(&self) -> &Mmr {
-        &self.0
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns a new [ChainMmr] instantiated from the provided partial MMR and a list of
+    /// (block_num, block_hash) tuples.
+    ///
+    /// # Panics
+    /// Panics if:
+    /// - The number of provided blocks is greater than the number of blocks in the chain.
+    /// - Any of the specified block numbers is greater than the chain height.
+    /// - The list of blocks contains duplicates.
```

Comment:
Can we return errors here instead of panicking?

Reply:
Just noticed the [`TODO`](https://github.com/0xPolygonMiden/miden-base/blob/123a9d4ff836d9162dca4b455dd2e44515dd22a9/objects/src/chain/mod.rs#L41); it's probably best to do it now, since changing it will result in a breaking change

Reply:
Agreed! Added `ChainMmrError`.